state/state.go

```diff
@@ -34,6 +40,110 @@ type DownloadedObject struct {
 	Object, Filename string
 }
 
+type DynamoDBStater struct {
+	*sync.Mutex
+	Session *session.Session
+	Service string
+}
+
+func NewDynamoDBStater(session *session.Session, service string) *DynamoDBStater {
+	return &DynamoDBStater{
+		Mutex:   &sync.Mutex{},
+		Session: session,
+		Service: service,
+	}
+}
+
+// list of processed objects
+func (d *DynamoDBStater) processedObjects() (map[string]time.Time, error) {
+	objs := make(map[string]time.Time)
+
+	type Record struct {
+		S3object string    `json:"S3object"`
+		Time     time.Time `json:"Time"`
+	}
+
+	var records []Record
+
+	svc := dynamodb.New(d.Session)
+	err := svc.ScanPages(&dynamodb.ScanInput{
+		TableName: aws.String("HoneyELBAccessLogBuckets"),
+	}, func(logs *dynamodb.ScanOutput, last bool) bool {
+		recs := []Record{}
+
+		err := dynamodbattribute.UnmarshalListOfMaps(logs.Items, &recs)
+
+		// break out of function
+		if err != nil {
+			panic(fmt.Sprintf("failed to unmarshal Dynamodb Scan Items, %v", err))
+		}
+
+		records = append(records, recs...)
+
+		return true
```

Comment:
Isn't this supposed to do something with `last` and return true/false depneding on whether there are more records to read? If it's always `return true` when does this loop end? Or is `return true` mean it's not reading any followups if there are more rows? If so, in the function definition, I think you can make `last` an unnamed variable (`_`)

Reply:
It'd be surprising if they looped forever if the function returns true, but to be clear I'd be inclined to return `!lastPage` here

Reply:
unfortunately, you can't insert map values in to dynamodb, the way they suggest to structure "map" values is to have a primary key value and then a sort key for that primary key
in this case, we use `S3Object` as the pk and `Time` as the sort key - and although `maps[string]time.Time` contains that info, it still needs to be marshaled into a struct for putItem
we could have it where `Time` is the pk, because that will be a unique value, but then we lose the string w/ the S3Object info in it

Reply:
Cool, we can leave as-is then.